File: src/mlip_plot/utils/water.py
"""
Water molecule identification and property extraction.

This module provides utilities for identifying water molecules in MD trajectories
and extracting their geometric properties (positions, distances, angles).
Supports periodic boundary conditions via minimum image convention.
"""

from dataclasses import dataclass
from typing import List, Dict, Any, Optional, Tuple
import numpy as np

# Import C++ core (required)
try:
    from mlip_plot._core import identify_water_molecules as _cpp_identify_water
    from mlip_plot._core import extract_water_properties as _cpp_extract_properties
except ImportError:
    raise ImportError("C++ core module not available. Please rebuild the package.")


@dataclass(frozen=True, slots=True)
class WaterMolecule:
    """
    Represents a water molecule with its geometric properties.

    Attributes
    ----------
    O_index : int
        Index of the oxygen atom in the frame
    H1_index : int
        Index of the first hydrogen atom
    H2_index : int
        Index of the second hydrogen atom
    O_position : np.ndarray
        Position of oxygen atom [x, y, z]
    H1_position : np.ndarray
        Position of first hydrogen [x, y, z]
    H2_position : np.ndarray
        Position of second hydrogen [x, y, z]
    OH1_distance : float
        O-H1 bond distance in Angstroms
    OH2_distance : float
        O-H2 bond distance in Angstroms
    H1H2_distance : float
        H1-H2 distance in Angstroms
    H1OH2_angle : float
        H1-O-H2 bond angle in degrees
    """
    O_index: int
    H1_index: int
    H2_index: int
    O_position: np.ndarray
    H1_position: np.ndarray
    H2_position: np.ndarray
    OH1_distance: float
    OH2_distance: float
    H1H2_distance: float
    H1OH2_angle: float

    def __repr__(self) -> str:
        return (f"WaterMolecule(O={self.O_index}, H1={self.H1_index}, H2={self.H2_index}, "
                f"angle={self.H1OH2_angle:.1f}deg)")


def _get_box_lengths(frames: List[Dict[str, Any]]) -> Tuple[float, float, float]:
    """Extract box lengths from frame data."""
    box = frames[0]['box']
    Lx = box['xhi'] - box['xlo']
    Ly = box['yhi'] - box['ylo']
    Lz = box['zhi'] - box['zlo']
    return (Lx, Ly, Lz)


def identify_water_molecules(
    frames: List[Dict[str, Any]],
    o_h_cutoff: float = 1.2,
) -> Tuple[np.ndarray, int]:
    """
    Identify water molecules from trajectory frames.

    Uses the first frame to identify O-H bonding topology, which is
    then assumed to remain constant throughout the trajectory.
    Supports periodic boundary conditions via minimum image convention.

    Parameters
    ----------
    frames : list of dict
        Trajectory frames from read_lammpstrj
    o_h_cutoff : float, optional
        O-H bond distance cutoff in Angstroms (default: 1.2)

    Returns
    -------
    molecule_indices : ndarray, shape (n_molecules, 3)
        Array of [O_index, H1_index, H2_index] for each water molecule
    n_molecules : int
        Number of water molecules found

    Raises
    ------
    ValueError
        If frames don't contain element information or no water molecules found
    """
    if not frames:
        raise ValueError("No frames provided")

    if frames[0]['elements'] is None:
        raise ValueError("Frames must contain element information")

    # Build element-to-type mapping
    elements = frames[0]['elements']
    types = frames[0]['types']
    positions = frames[0]['positions']

    # Find O and H type indices
    o_type_idx = None
    h_type_idx = None

    for elem, typ in zip(elements, types):
        if elem == 'O' and o_type_idx is None:
            o_type_idx = typ - 1  # 0-indexed
        elif elem == 'H' and h_type_idx is None:
            h_type_idx = typ - 1

        if o_type_idx is not None and h_type_idx is not None:
            break

    if o_type_idx is None:
        raise ValueError("No oxygen atoms found in trajectory")
    if h_type_idx is None:
        raise ValueError("No hydrogen atoms found in trajectory")

    # Convert types to 0-indexed
    element_types = (types - 1).astype(np.int32)

    # Get box dimensions for PBC
    box_lengths = _get_box_lengths(frames)

    # Use C++ backend
    molecule_indices, n_molecules = _cpp_identify_water(
        positions.astype(np.float64),
        element_types,
        o_type_idx,
        h_type_idx,
        box_lengths,
        o_h_cutoff
    )

    if n_molecules == 0:
        raise ValueError("No water molecules found")

    return molecule_indices, n_molecules




def extract_water_molecules(
    frames: List[Dict[str, Any]],
    o_h_cutoff: float = 1.2,
) -> List[List[WaterMolecule]]:
    """
    Extract water molecule properties for all frames.

    Each frame is identified independently to ensure correct O-H assignments
    even when atoms move significantly between frames.

    Supports periodic boundary conditions via minimum image convention
    for distance and angle calculations.

    Parameters
    ----------
    frames : list of dict
        Trajectory frames from read_lammpstrj
    o_h_cutoff : float, optional
        O-H bond cutoff for identification (default: 1.2)

    Returns
    -------
    water_molecules : list of list of WaterMolecule
        Nested list where water_molecules[frame_idx][mol_idx] gives
        the WaterMolecule object for that frame and molecule.

    Note
    ----
    Water molecules are identified independently for each frame, so
    water indices may not be consistent across frames (Water 0 in frame 0
    may be a different physical molecule than Water 0 in frame 1).

    Example
    -------
    >>> frames = read_lammpstrj('trajectory.lammpstrj')
    >>> waters = extract_water_molecules(frames)
    >>> # Get first water molecule in frame 0
    >>> mol = waters[0][0]
    >>> print(f"O-H distance: {mol.OH1_distance:.3f} A")
    >>> print(f"H-O-H angle: {mol.H1OH2_angle:.1f} degrees")
    """
    all_waters = []

    for frame in frames:
        # Identify water molecules for THIS frame
        molecule_indices, n_molecules = identify_water_molecules([frame], o_h_cutoff)

        if n_molecules == 0:
            all_waters.append([])
            continue

        # Get box dimensions for this frame
        box_lengths = _get_box_lengths([frame])

        # Use C++ backend for single frame
        positions_list = [frame['positions'].astype(np.float64)]

        props = _cpp_extract_properties(
            positions_list,
            molecule_indices.astype(np.int32),
            box_lengths
        )

        # Convert to WaterMolecule objects
        frame_waters = []
        for m in range(n_molecules):
            mol = WaterMolecule(
                O_index=int(molecule_indices[m, 0]),
                H1_index=int(molecule_indices[m, 1]),
                H2_index=int(molecule_indices[m, 2]),
                O_position=props['O_positions'][0, m].copy(),
                H1_position=props['H1_positions'][0, m].copy(),
                H2_position=props['H2_positions'][0, m].copy(),
                OH1_distance=float(props['OH1_distances'][0, m]),
                OH2_distance=float(props['OH2_distances'][0, m]),
                H1H2_distance=float(props['H1H2_distances'][0, m]),
                H1OH2_angle=float(props['H1OH2_angles'][0, m]),
            )
            frame_waters.append(mol)
        all_waters.append(frame_waters)

    return all_waters

File: src/cpp/src/water.cpp
/**
 * Water molecule identification and property extraction.
 *
 * Provides functions for identifying water molecules in MD trajectories
 * and computing their geometric properties (positions, distances, angles).
 * Supports periodic boundary conditions via minimum image convention.
 */

#include "mlip_core.hpp"
#include <cmath>
#include <algorithm>
#include <vector>
#include <unordered_set>

#ifdef USE_OPENMP
#include <omp.h>
#endif

namespace mlip {

namespace {

/**
 * Apply minimum image convention for periodic boundary conditions.
 * Modifies dx, dy, dz in place to be the shortest distance vector.
 */
inline void apply_minimum_image(
    double& dx, double& dy, double& dz,
    const std::array<double, 3>& box_lengths
) {
    if (box_lengths[0] > 0) {
        dx = dx - std::round(dx / box_lengths[0]) * box_lengths[0];
    }
    if (box_lengths[1] > 0) {
        dy = dy - std::round(dy / box_lengths[1]) * box_lengths[1];
    }
    if (box_lengths[2] > 0) {
        dz = dz - std::round(dz / box_lengths[2]) * box_lengths[2];
    }
}

/**
 * Calculate distance between two 3D points with PBC support.
 */
inline double distance_pbc(
    const double* pos1,
    const double* pos2,
    const std::array<double, 3>& box_lengths
) {
    double dx = pos1[0] - pos2[0];
    double dy = pos1[1] - pos2[1];
    double dz = pos1[2] - pos2[2];

    apply_minimum_image(dx, dy, dz, box_lengths);

    return std::sqrt(dx * dx + dy * dy + dz * dz);
}

/**
 * Calculate the displacement vector from pos1 to pos2 with PBC.
 * Returns the vector (pos2 - pos1) under minimum image convention.
 */
inline void displacement_pbc(
    const double* pos1,
    const double* pos2,
    const std::array<double, 3>& box_lengths,
    double* disp
) {
    disp[0] = pos2[0] - pos1[0];
    disp[1] = pos2[1] - pos1[1];
    disp[2] = pos2[2] - pos1[2];

    apply_minimum_image(disp[0], disp[1], disp[2], box_lengths);
}

/**
 * Calculate angle at vertex B for triangle A-B-C with PBC support.
 * Returns angle in degrees.
 *
 * Uses the dot product formula:
 *   cos(angle) = (BA . BC) / (|BA| * |BC|)
 */
inline double angle_abc_pbc(
    const double* A,
    const double* B,
    const double* C,
    const std::array<double, 3>& box_lengths
) {
    // Vectors BA and BC with PBC
    double ba[3], bc[3];
    displacement_pbc(B, A, box_lengths, ba);
    displacement_pbc(B, C, box_lengths, bc);

    double dot = ba[0] * bc[0] + ba[1] * bc[1] + ba[2] * bc[2];
    double mag_ba = std::sqrt(ba[0] * ba[0] + ba[1] * ba[1] + ba[2] * ba[2]);
    double mag_bc = std::sqrt(bc[0] * bc[0] + bc[1] * bc[1] + bc[2] * bc[2]);

    if (mag_ba < 1e-10 || mag_bc < 1e-10) {
        return 0.0;
    }

    double cos_angle = dot / (mag_ba * mag_bc);
    // Clamp for numerical safety
    cos_angle = std::max(-1.0, std::min(1.0, cos_angle));

    // Convert to degrees
    return std::acos(cos_angle) * 180.0 / M_PI;
}

}  // anonymous namespace

WaterIdentificationResult identify_water_molecules(
    const double* positions,
    const int* element_types,
    size_t n_atoms,
    int o_type_idx,
    int h_type_idx,
    const std::array<double, 3>& box_lengths,
    double o_h_cutoff
) {
    WaterIdentificationResult result;

    // Collect O and H atom indices
    std::vector<int> o_indices;
    std::vector<int> h_indices;

    for (size_t i = 0; i < n_atoms; ++i) {
        if (element_types[i] == o_type_idx) {
            o_indices.push_back(static_cast<int>(i));
        } else if (element_types[i] == h_type_idx) {
            h_indices.push_back(static_cast<int>(i));
        }
    }

    if (o_indices.empty() || h_indices.size() < 2) {
        result.n_molecules = 0;
        return result;
    }

    std::unordered_set<int> used_h;

    // For each O atom, find the two closest H atoms within cutoff
    for (int o_idx : o_indices) {
        const double* o_pos = positions + o_idx * 3;

        // Find H atoms within cutoff
        std::vector<std::pair<int, double>> bonded_h;

        for (int h_idx : h_indices) {
            if (used_h.count(h_idx)) continue;

            const double* h_pos = positions + h_idx * 3;
            double dist = distance_pbc(o_pos, h_pos, box_lengths);

            if (dist < o_h_cutoff) {
                bonded_h.push_back({h_idx, dist});
            }
        }

        // Sort by distance and take closest 2
        if (bonded_h.size() >= 2) {
            std::sort(bonded_h.begin(), bonded_h.end(),
                [](const auto& a, const auto& b) { return a.second < b.second; });

            int h1_idx = bonded_h[0].first;
            int h2_idx = bonded_h[1].first;

            result.molecule_indices.push_back({o_idx, h1_idx, h2_idx});
            used_h.insert(h1_idx);
            used_h.insert(h2_idx);
        }
    }

    result.n_molecules = result.molecule_indices.size();
    return result;
}

std::vector<WaterMolecule> extract_water_properties(
    const double* positions,
    const std::vector<std::array<int, 3>>& molecule_indices,
    size_t n_molecules,
    const std::array<double, 3>& box_lengths
) {
    std::vector<WaterMolecule> molecules(n_molecules);

#ifdef USE_OPENMP
    #pragma omp parallel for schedule(static)
#endif
    for (size_t i = 0; i < n_molecules; ++i) {
        const auto& indices = molecule_indices[i];
        WaterMolecule& mol = molecules[i];

        mol.O_index = indices[0];
        mol.H1_index = indices[1];
        mol.H2_index = indices[2];

        const double* o_pos = positions + indices[0] * 3;
        const double* h1_pos = positions + indices[1] * 3;
        const double* h2_pos = positions + indices[2] * 3;

        // Copy positions (raw positions, not unwrapped)
        for (int j = 0; j < 3; ++j) {
            mol.O_position[j] = o_pos[j];
            mol.H1_position[j] = h1_pos[j];
            mol.H2_position[j] = h2_pos[j];
        }

        // Calculate distances with PBC
        mol.OH1_distance = distance_pbc(o_pos, h1_pos, box_lengths);
        mol.OH2_distance = distance_pbc(o_pos, h2_pos, box_lengths);
        mol.H1H2_distance = distance_pbc(h1_pos, h2_pos, box_lengths);

        // Calculate H1-O-H2 angle (angle at O between the two H atoms) with PBC
        mol.H1OH2_angle = angle_abc_pbc(h1_pos, o_pos, h2_pos, box_lengths);
    }

    return molecules;
}

std::vector<std::vector<WaterMolecule>> extract_water_properties_multiframe(
    const std::vector<const double*>& all_positions,
    const std::vector<std::array<int, 3>>& molecule_indices,
    size_t n_molecules,
    size_t n_frames,
    const std::array<double, 3>& box_lengths
) {
    std::vector<std::vector<WaterMolecule>> all_molecules(n_frames);

#ifdef USE_OPENMP
    #pragma omp parallel for schedule(dynamic)
#endif
    for (size_t f = 0; f < n_frames; ++f) {
        all_molecules[f] = extract_water_properties(
            all_positions[f], molecule_indices, n_molecules, box_lengths
        );
    }

    return all_molecules;
}

}  // namespace mlip
